File: day23/part2.py
from functools import cache

grid = open(0).read().splitlines()
dest = (140, 139)


def valid(r, c):
    return (
        r >= 0
        and r < len(grid)
        and c >= 0
        and c < len(grid[0])
        and grid[r][c] != "#"
    )


def adjacent(r, c):
    return [(r + 1, c), (r - 1, c), (r, c + 1), (r, c - 1)]


@cache
def distance(p, q):
    queue = [(0, *p)]
    seen = set()
    p_neighbors = neighbors(*p)
    d_max = 0
    while queue:
        d, r, c = queue.pop()

        if (r, c) == q:
            d_max = max(d_max, d)

        if (r, c) in seen or (r, c) in p_neighbors:
            continue
        seen.add((r, c))

        for nr, nc in adjacent(r, c):
            if valid(nr, nc):
                queue.append((d + 1, nr, nc))
    return d_max


@cache
def neighbors(r0, c0):
    queue = [(r0, c0)]
    seen = set()
    res = []
    while queue:
        r, c = queue.pop()

        if (r, c) in seen:
            continue
        seen.add((r, c))

        if (r, c) == dest:
            res.append((r, c))
            continue

        passages = 0
        if (r, c) != (r0, c0):
            for nr, nc in adjacent(r, c):
                if valid(nr, nc):
                    passages += 1
        if passages >= 3:
            res.append((r, c))
            continue

        for nr, nc in adjacent(r, c):
            if valid(nr, nc) or (nr, nc) == dest:
                queue.append((nr, nc))
    return res


seen = [[False for _ in range(len(grid[0]))] for _ in range(len(grid))]
answer = 0


def dfs(r, c, d):
    global answer

    if (r, c) == dest:
        answer = max(answer, d)
        return

    if seen[r][c]:
        return

    seen[r][c] = True

    for nr, nc in neighbors(r, c):
        nd = d + distance((r, c), (nr, nc))
        dfs(nr, nc, nd)

    seen[r][c] = False


dfs(0, 1, 0)
print(answer)

File: day23/part2.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define LEN(x) (sizeof(x) / sizeof((x)[0]))
#define DIFF ((int[4]){cols, -cols, 1, -1})

char grid[65536];
int rows, cols, goal;

bool
valid(int p)
{
	return p >= 0 && p < rows * cols && grid[p] != '#';
}

bool
passage(int p)
{
	if (p == goal)
		return true;

	int a = 0;
	for (int i = 0; i < 4; i++) {
		int q = p + DIFF[i];
		a += valid(q);
	}
	return a > 2;
}

struct distmap_node {
	struct distmap_node *next;
	int p, q;
	int dist;
};

struct distmap_node *distmap[1009] = {0};

int
distance(int p, int q)
{
	int hash = (p + q * 3) % LEN(distmap);
	struct distmap_node *dmn;
	for (dmn = distmap[hash]; dmn != NULL; dmn = dmn->next)
		if (dmn->p == p && dmn->q == q)
			return dmn->dist;

	struct state {
		int pos;
		int dist;
	} stack[1024] = {{p, 0}};

	bool seen[LEN(grid)] = {0};
	int maxdist = 0;
	int k = 1;
	while (k > 0) {
		assert(k < LEN(stack));

		k--;
		int r = stack[k].pos;
		int dist = stack[k].dist;

		if (r == q) {
			if (dist > maxdist)
				maxdist = dist;
			continue;
		}

		if (r != p && passage(r))
			continue;

		if (seen[r])
			continue;
		seen[r] = true;

		for (int i = 0; i < 4; i++) {
			int s = r + DIFF[i];
			if (valid(s)) {
				stack[k].pos = s;
				stack[k].dist = dist + 1;
				k++;
			}
		}
	}

	dmn = malloc(sizeof(*dmn));
	dmn->p = p;
	dmn->q = q;
	dmn->dist = maxdist;
	dmn->next = distmap[hash];
	distmap[hash] = dmn;
	return maxdist;
}

int
neighbors(int p, int **res)
{
	static int cache[LEN(grid)][5];

	if (*cache[p])
		goto ret;

	bool seen[LEN(grid)] = {0};
	int stack[1024] = {p};
	int k = 1;
	int n = 0;
	while (k > 0) {
		assert(k < LEN(stack));

		k--;
		int q = stack[k];

		if (seen[q])
			continue;
		seen[q] = true;

		if (q != p && passage(q)) {
			cache[p][++n] = q;
			continue;
		}

		for (int i = 0; i < 4; i++) {
			int r = q + DIFF[i];
			if (valid(r))
				stack[k++] = r;
		}
	}
	*cache[p] = n;
ret:
	*res = cache[p] + 1;
	return *cache[p];
}

int answer = 0;

void
search(int p, int d)
{
	static bool seen[LEN(grid)];

	if (p == goal) {
		if (d > answer)
			answer = d;
		return;
	}

	if (seen[p])
		return;
	seen[p] = true;

	int *adj;
	int n = neighbors(p, &adj);
	for (int i = 0; i < n; i++) {
		int q = adj[i];
		search(q, d + distance(p, q));
	}
	seen[p] = false;
}

void
init(void)
{
	int i = 0, r = 0, c;
	while ((c = getchar()) != EOF)
		if (c == '\n')
			r++;
		else
			grid[i++] = c;
	rows = r;
	cols = i / r;
	goal = rows * cols - 2;
}

int
main(void)
{
	init();
	search(1, 0);
	printf("%d\n", answer);
	return 0;
}
